download-utils/src/lib.rs

```diff
@@ -84,38 +132,83 @@ pub fn download_file(
         info!("Downloading {} bytes from {}", download_size, url);
     }
 
-    struct DownloadProgress<R> {
+    struct DownloadProgress<R, F>
+    where
+        F: Fn(&DownloadProgressRecord) -> bool,
+    {
         progress_bar: ProgressBar,
         response: R,
         last_print: Instant,
         current_bytes: usize,
         last_print_bytes: usize,
         download_size: f32,
         use_progress_bar: bool,
+        start_time: Instant,
+        callback: Option<F>,
+        notification_count: u64,
     }
 
-    impl<R: Read> Read for DownloadProgress<R> {
+    impl<R: Read, F> Read for DownloadProgress<R, F>
+    where
+        F: Fn(&DownloadProgressRecord) -> bool,
+    {
         fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-            self.response.read(buf).map(|n| {
-                if self.use_progress_bar {
-                    self.progress_bar.inc(n as u64);
-                } else {
+            match self.response.read(buf) {
```

Comment:
consider this:

https://github.com/jeffwashington/solana/blob/e4861e0435dc604e305e0822b0915bd4dee845a2/download-utils/src/lib.rs#L156-L210

Reply:
Yes. Looks cleaner!